src/cns.ts

```diff
@@ -0,0 +1,219 @@
+   */
+  async address(domain: string, currencyTicker: string): Promise<string> {
+    const [nodeHash, _, __, resolver] = await this.getResolutionMeta(domain);
+    const addr: string = await this.fetchAddress(resolver, nodeHash, currencyTicker);
+    console.log({addr});
+    
+    if (!addr)
+      throw new ResolutionError(ResolutionErrorCode.UnspecifiedCurrency, {
+        domain,
+        currencyTicker,
+      });
+    return addr;
+  }
+
+   /**
+   * @ignore
+   * @param resolver - Resolver address
+   * @param nodeHash - namehash of a domain name
+   */
+  private async fetchAddress(resolver: string, nodeHash: string, coinName?: string): Promise<string> {
+    console.log('loool');
+    const resolverContract = new Contract(
+      this.url,
+      resolverInterface,
+      resolver
+    );
+    const addrKey = `crypto.${coinName.toUpperCase()}.address`;
+    console.log({addrKey});
+    const addr: string = await this.callMethod(resolverContract, 'get', [addrKey, nodeHash]);
+    console.log({addr});
+    return addr;
+  }
+
+  /** @ignore */
+  private async getResolver(nodeHash): Promise<string> {
+    
+    const test =  await this.callMethod(this.cnsContract, 'resolverOf', [nodeHash]);
+    console.log('getting resolver');
+    console.log({resolver: test});
+    return test;
+    
+  }
+
+  /** @ignore */
+  async owner(nodeHash): Promise<string> {
+    return await this.callMethod(this.cnsContract, 'ownerOf', [nodeHash]);
+  }
+
+  /** @internal */
+  async record(domain: string, key: string): Promise<string> {
+    // TODO: implement record
+    throw new Error("Method not implemented.");
+  }  
+
+  /**
+   * @ignore
+   * This was done to make automated tests more configurable
+   * @param domain
+   * @retuns Promise that resolves to [nodehash, owner, ttl, resolver]
+   */
+  private async getResolutionMeta(domain: string): Promise<[string, string, number, string]> {
+    
+    const nodeHash = this.namehash(domain);
+    const ownerPromise = this.owner(nodeHash);
+    
+    
+    const resolver:string = await this.getResolver(nodeHash);
+    
+    if (!resolver || resolver === NullAddress) {
+      var owner = await ownerPromise;
+      if (!owner || owner === NullAddress)
+        throw new ResolutionError(ResolutionErrorCode.UnregisteredDomain, { domain });
+      throw new ResolutionError(ResolutionErrorCode.UnspecifiedResolver, { domain });
+    }
+    const resolverContract = new Contract(this.url, resolverInterface, resolver);
+    // const ttl = await this.callMethod(resolverContract, 'get', ['ttl', nodeHash]);
+    const ttl = '0';
+    console.log({ttl});
+    
+    return [nodeHash, owner, parseInt(ttl) || 0, resolver];
+  }
+
+   /**
+   * Internal wrapper for ens method. Used to throw an error when ens is down
+   *  @param method - method to be called
+   *  @throws ResolutionError -> When blockchain is down
+   */
+  private async callMethod(contract: Contract, methodname: string, params: any
```

Comment:
This method looks the same for ENS and CNS too. Lets extract it. The only differentiation i `method: "CNS"` which can be done by having a property with this name for each service like 

class NamingService {
   readonly name: string;
}

class Ens extends NamingService {
  readonly name = "ENS";
}
```